src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+      }
+
+      // now check to see if offsetParent is a fixed layer
+      if (el === op) {
+        if (computedStyle(win, op).position == 'fixed') {
+          LayoutLayers.declareLayer(op);
+          return op;
+        }
+        op = op.offsetParent;
+      }
+
+      last = el;
+    }
+
+
+    if (last.nodeType !== Node.DOCUMENT_NODE) {
+      throw dev().createError('element is not in the DOM tree');
+    }
+    return null;
+  }
+
+  contains(element) {
+    const root = this.element_;
+    return root !== element && root.contains(element);
+  }
+
+  add(child) {
+    dev().assert(this.isLayer());
+    dev().assert(this.contains(child));
+
+    const layout = LayoutElement.for(child);
+    this.children_.push(layout);
+  }
+
+  remove(child) {
+    dev().assert(this.isLayer());
+
+    const layout = LayoutElement.for(child);
+    dev().assert(layout.getParentLayer() === this);
+
+    layout.undeclareLayer();
+    layout.forgetParentLayer();
+    const i = this.children_.indexOf(layout);
+    if (i > -1) {
+      this.children_.splice(i, 1);
+    }
+  }
+
+  isLayer() {
+    return this.isLayer_;
+  }
+
+  declareLayer() {
+    this.isLayer_ = true;
+    this.needsRemeasure_ = true;
+    this.needsScrollRemeasure_ = true;
+
+    const element = this.element_;
+    const scroller = getScrollingElement(element);
+    this.isRootLayer_ = element === scroller || !scroller.contains(element);
+
+    const parent = this.getParentLayer();
+    if (parent) {
+      parent.transfer_(this);
+    }
+  }
+
+  /**
+   * This is necessary when the screen size changes (or other similar events
+   * happen), because CSS styles may change. We'll need to recompute layers
+   * lazily after this happens.
+   */
+  undeclareLayer() {
+    if (!this.isLayer_ || this.isRootLayer_) {
+      return;
+    }
+
+    this.isLayer_ = false;
+    this.transfer_(this.getParentLayer());
+  }
+
+  transfer_(layer) {
+    // An optimization if we know that the new layer definitely contains
+    // everything in this layer.
+    const contained = layer.contains(this.element_);
+
+    filterSplice(this.children_, layout => {
+      if (contained || layer.contains(layout.element_)) {
+        layout.parentLayer_ = layer;
+        layer.children_.push(layout);
+        return false;
+      }
+
+      return true;
+    });
+  }
+
+  /**
+   * Gets the element's parent layer. If this element is itself a layer, it
+   * returns the layer's parent.
+   * @return {?LayoutElement}
+   */
+  getParentLayer() {
+    if (this.parentLayer_ === undefined) {
+      const parent = LayoutElement.getParentLayer(this.element_, true);
+      this.parentLayer_ = parent;
+      if (parent) {
+        parent.add(this.element_);
+      }
+    }
+    return this.parentLayer_;
+  }
+
+  forgetParentLayer() {
+    this.parentLayer_ = undefined;
+  }
+
+  /**
+   * Gets the size of the element.
+   * @return {!SizeDef}
+   */
+  getSize() {
+    if (this.needsRemeasure_) {
+      this.remeasure();
```

Comment:
Does this mean that this call always requires vsync?

Reply:
Yes and no. This was an optimization I added because doing the initial measurements during startup was causing noticeable lag.

I'd imagined the large majority of cases, this value will be cached. Only in cases where they're doing a mutate followed immediately (before the next Resources Pass) by a measure will it need to actually measure.